python/perf-kernels/flash-attention.py

```diff
@@ -544,132 +591,198 @@ def _bwd_kernel_dk_dv(
         order=(1, 0)
     )
+    off_zh = off_z * num_h + off_h * 1
     # pointer to row-wise quantities in value-like data
-    D_ptrs = D + off_hz * seqlen_q
-    l_ptrs = L + off_hz * seqlen_q
-    qk_scale = sm_scale * 1.44269504
+    D_ptrs = D + off_zh * seqlen_q
+    l_ptrs = L + off_zh * seqlen_q
+    qk_scale = sm_scale * 1.44269504089
     # load k and v: they will stay in SRAM throughout
-    k = tl.load(K_block_ptr)
-    k = (k * qk_scale).to(K_block_ptr.type.element_ty)
-    vt = tl.load(VT_block_ptr)
+    # (BLOCK_DMODEL, BLOCK_N)
+    if PADDED_HEAD:
+        kt = tl.load(KT_block_ptr, boundary_check=(1,0), padding_option="zero")
+    else:
+        kt = tl.load(KT_block_ptr, boundary_check=(1,), padding_option="zero")
+    kt = (kt * qk_scale).to(KT_block_ptr.type.element_ty)
+    # (BLOCK_DMODEL, BLOCK_N)
+    if PADDED_HEAD:
+        vt = tl.load(VT_block_ptr, boundary_check=(1,0), padding_option="zero")
+    else:
+        vt = tl.load(VT_block_ptr, boundary_check=(1,), padding_option="zero")
     dv = tl.zeros([BLOCK_N, BLOCK_DMODEL], dtype=tl.float32)
     dk = tl.zeros([BLOCK_N, BLOCK_DMODEL], dtype=tl.float32)
+
     # This lower loop bound is because of the causal mask. We create a lower triangular
     # result. The upper triangular is -inf (becomes 0 when we do e^x). As such, it can
     # be ignored in the GEMM.
-    lo = start_m if CAUSAL else 0
+    lo = (start_m // BLOCK_M) * BLOCK_M if CAUSAL else 0
     hi = seqlen_q
     Q_block_ptr = tl.advance(Q_block_ptr, (lo, 0))
     DO_block_ptr = tl.advance(DO_block_ptr, (lo, 0))
-    batch_philox_offset = philox_offset_base + off_hz * seqlen_q * seqlen_k
+    batch_philox_offset = philox_offset_base + off_zh * seqlen_q * seqlen_k
+    '''
+           K1   K2      (d)V      dO
+    Q1    qk11 qk12     (d)v1     dO1
+    Q2    qk21 qk22     (d)v2     dO2
+
+    QK: (seqlen_q, seqlen_k)
+    dO: (seqlen_q, hdim)
+    dV: (seqlen_k, hdim)
+
+    dV = (QK)^T dO
+
+    dV1 = qk11 dO1 + qk21 dO2 = q1 k1 dO1 + q2 k1 dO2
+    dV2 = qk12 dO1 + qk22 dO2 = q1 k2 dO1 + q2 k2 dO2
+                                ~~~~~ = 0
+    start_m: select k and dV
+    start_n: select q and dO
+    '''
     # loop over q (seqlen_q, dhead), do (seqlen_q, d_head)
     for start_n in range(lo, hi, BLOCK_M):
-        offs_m_curr = offs_n[:, None] + start_n
+        offs_m_curr = offs_n[:, None] + start_n # (BLOCK_M, 1)
         # -- load q, do --
-        q = tl.load(Q_block_ptr)
-        do = tl.load(DO_block_ptr)
+        # TODO: It is more optimal to do OOB check only in the last iter.
+        # (BLOCK_M, BLOCK_DMODEL), offs = (BLOCK_M * iter, 0) = (start_n, 0)
+        # do is (BLOCK_M, BLOCK_DMODEL)
+        if PADDED_HEAD:
+            q = tl.load(Q_block_ptr, boundary_check=(0,1), padding_option="zero")
+            do = tl.load(DO_block_ptr, boundary_check=(0,1), padding_option="zero")
+        else:
+            q = tl.load(Q_block_ptr, boundary_check=(0,), padding_option="zero")
```

Comment:
Needs one more level of branching.
However if we are aiming for the performance we should consider commenting out the `boundary_check` for now